### DISCUSSION TITLE
Why does tamagui require react-native with Next.js?

### DISCUSSION THREAD

**Discussion by cseas** [tamagui/tamagui#649] [General]
The [Next.js section in the docs][1] mentions the below import. 

```tsx
import { AppRegistry } from 'react-native'
```

Next.js is used to build web applications. Why does tamagui require web projects to install react-native as a dependency?

Is it because the library is based on react-native-web? The [Introduction][2] page doesn't mention anything about this so I'm confused.

As someone who exclusively develops web apps, I've never needed to use react-native or react-native-web in my projects so seeing an import from react-native required in a web project just looks wrong. I shouldn't need to ship any react-native dependencies to browsers as a web developer.

It makes sense for React Native developers since their priority is API consistency across different codebases. But for someone who's only shipping a web app, I shouldn't need to ship any dependencies that I don't really need.

Getting clarity on this would help evaluate the decision to use tamagui in projects vs other libraries. From the looks of it, it seems tamagui makes sense for React Native devs who want to ship web apps. But for pure web apps, some other library would probably be a better choice?

[1]: https://tamagui.dev/docs/guides/next-js
[2]: https://tamagui.dev/docs/intro/introduction

**Comment by redbar0n** [comment]
I agree that it is confusing, and that the [_document.tsx](https://tamagui.dev/docs/guides/next-js#_documenttsx) section in the Next.js guide should document it better.

Here's an explanation that I hope makes sense:

Tamagui builds on React Native for Web. It is the API that enables the cross-platform interaction, through rendering React Native components on the web (by turning `<View>` into `<div>` etc.). 

If you see the examples even on the [the React Native for Web page](https://necolas.github.io/react-native-web/), they import from 'react-native', but they alias it to react-native-web when building the app for the web:

> The idea is that you write code importing from React-Native (so that it will run on Android / iOS), but when you're building your app for the browser, you alias the react-native module to instead point at the react-native-web module.

https://github.com/necolas/react-native-web/issues/424

> But for someone who's only shipping a web app, I shouldn't need to ship any dependencies that I don't really need.

In Tamagui you can slim down React Native for Web by aliasing it further to use Nate's own Lite version called [react-native-web-lite](https://github.com/tamagui/tamagui/tree/master/packages/react-native-web-lite).

**Reply by cseas** [reply]
Thanks! That's the clarification I was looking for. I couldn't find any references in the docs that mention tamagui is based on React Native for Web instead of React.

I've never personally worked on React Native or React Native Web so correct me if I'm wrong, but looks like the reasons for choosing React Native Web over React are [somewhat aesthetic][1]. If I'm only developing for web, choosing React Native Web or anything based on it wouldn't really make sense.

So tamagui is probably not the best choice for pure web projects since web developers don't really need the React Native Web abstraction it builds upon. Would love to see a section in the docs explaining this if tamagui does outperform other pure web libraries.

[1]: https://stackoverflow.com/a/56554898/7435656

**Reply by redbar0n** [reply]
Yes, Tamagui currently isn't the _obvious_ choice for web, since it has so many great counterparts on the web (Chakra, or Tailwind/DaisyUI, for instance). And although Tamagui is even more powerful than them, they have more components.

I've made some potentially helpful notes on React Native for Web (option 3) in this article, that may guide the decision to choose it or not: [The different strategies to building a cross platform app](https://dev.to/this-is-learning/the-different-strategies-to-building-a-cross-platform-app-3p56).

The greatest benefit with using Tamagui when building a web project isn't aesthetical (although the API is great, and [RNW takes away  some CSS footguns](https://dev.to/redbar0n/how-to-use-chakraui-on-native-mobile-53hg)), but that [a later path](https://mxstbr.com/thoughts/tech-choice-regrets-at-spectrum/) to a React Native app, if you or your users ever want that, is next to nothing, since it's just a matter of compilation target. But it may not be a benefit, if you'd never need a native app.

That said, Tamagui is the closest you'd get to Facebook's own (yet unreleased, but [likely to be released?](https://stylex-docusaurus-nmn.vercel.app/docs/)) StyleX styling solution ([according to Naman Goel](https://twitter.com/naman34/status/1599581414487359489), the StyleX lead developer), which is built for scale and solves [many](https://www.youtube.com/watch?v=tFFn39lLO-U) of the [problems](https://www.youtube.com/watch?v=ur-sGzUWId4&t=157s) that come [with](https://twitter.com/necolas/status/1058949372837122048?s=20&t=joLGZLea9XAX4Eq-LtOJEA) CSS at scale.

Tamagui is also one of the [very few](https://andreipfeiffer.dev/blog/2021/css-in-js-styles-output) CSS-in-JS solutions (apart from [Sprinkles](https://vanilla-extract.style/documentation/packages/sprinkles/)) to my knowledge that gives you dynamic styling (style ternaries) while also doing compile-time atomic CSS extraction. That means it's very good for [SSR and responsivity/media-queries](https://dev.to/redbar0n/how-to-combine-react-native-web-responsivity-nextjs-ssr-to-get-seo-a4g), and doesn't force a client to execute JS just to apply some CSS (which is downloaded in parallel) to the incoming SSR'ed HTML.

Regarding performance, Tamagui's [web benchmark](https://tamagui.dev/docs/intro/benchmarks#web) compares it to some popular crossplatform but also web specific styling solutions.

**Reply by natew** [reply]
Actually Tamagui used to be built on top of RNW, but as of today it doesn't use RNW at all it just supports the superset of the API surface. But if you do use RNW with it, you need that AppRegistry stuff. So I'll update the docs and clarify that a bit.

**Reply by natew** [reply]
If you are using `tamagui` it uses more RN features and then pulls in RNW. But `@tamagui/core` doesn't need anything but itself and React to run on web/native.

**Reply by natew** [reply]
Added some notes here: 8592147127fadf723421e12a2ea4e41ed6a93145

**Reply by natew** [reply]
Not true - core and web both can be used without RNW still. We only deprecated `react-native-web-lite` our fork since rnw now has some improvements and we don't want to maintain a fork.

**Reply by redbar0n** [reply]
Thanks for correcting (deleted my comment then). I found out when I dug into the v 1.75.0 commit.

It’s been a bit confusing how much Tamagui relies on RNW. Re the v1.0 blog post.

If I understand you correctly and the commit you referenced above, then Tamagui now only relies on RNW for Image and Input components, but could be used on either native (using only core), or web, if you don’t use either of those components. Please correct me if I’m wrong.

I base this of the commit you referenced above (from Feb 2023):

> Note that we only are importing `react-native` here because if you use `tamagui` views like Input or Image they rely on React Native Image/Input (and therefore on the web, `react-native-web`). If you are just using core, or aren't using Image or Input in `tamagui`, you can forego the entire `AppRegistry` import and setup here. We're working on making even `tamagui` completely de-coupled from RN/RNW by building their own image and input components eventually.
